openmls/tests/test_group.rs

```diff
@@ -763,3 +763,257 @@ fn group_operations() {
+        )
+            .unwrap();
+
+        // Mandatory extensions
+        let capabilities_extension = Extension::Capabilities(CapabilitiesExtension::new(
+            None,
+            Some(&[ciphersuite.name()]),
+            None,
+        ));
+        let lifetime_extension = Extension::LifeTime(LifetimeExtension::new(60));
+        let mandatory_extensions: Vec<Extension> = vec![capabilities_extension, lifetime_extension];
+
+        // Generate KeyPackages
+        let alice_key_package_bundle = KeyPackageBundle::new(
+            &[ciphersuite.name()],
+            &alice_credential_bundle,
+            mandatory_extensions.clone(),
+        )
+            .unwrap();
+
+        let bob_key_package_bundle = KeyPackageBundle::new(
+            &[ciphersuite.name()],
+            &bob_credential_bundle,
+            mandatory_extensions.clone(),
+        )
+            .unwrap();
+        let bob_key_package = bob_key_package_bundle.key_package();
+
+        // === Alice creates a group ===
+        let group_id = [1, 2, 3, 4];
+        let mut group_alice = MlsGroup::new(
+            &group_id,
+            ciphersuite.name(),
+            alice_key_package_bundle,
+            MlsGroupConfig::default(),
+            None, /* Initial PSK */
+            None, /* MLS version */
+        )
+            .expect("Could not create group.");
+
+        // === Alice adds Bob ===
+        let bob_add_proposal = group_alice
+            .create_add_proposal(group_aad, &alice_credential_bundle, bob_key_package.clone())
+            .expect("Could not create proposal.");
+        let epoch_proposals = &[&bob_add_proposal];
+        let (mls_plaintext_commit, welcome_bundle_alice_bob_option, kpb_option) = group_alice
+            .create_commit(
+                group_aad,
+                &alice_credential_bundle,
+                epoch_proposals,
+                &[],
+                false,
+                None, /* PSK fetcher */
+            )
+            .expect("Error creating commit");
+        let commit = match mls_plaintext_commit.content() {
+            MlsPlaintextContentType::Commit(commit) => commit,
+            _ => panic!("Wrong content type"),
+        };
+        group_alice
+            .apply_commit(
+                &mls_plaintext_commit,
+                epoch_proposals,
+                &[],
+                None, /* PSK fetcher */
+            )
+            .expect("error applying commit");
+        let ratchet_tree = group_alice.tree().public_key_tree_copy();
+
+        let mut group_bob = match MlsGroup::new_from_welcome(
+            welcome_bundle_alice_bob_option.unwrap(),
+            Some(ratchet_tree),
+            bob_key_package_bundle,
+            None, /* PSK fetcher */
+        ) {
+            Ok(group) => group,
+            Err(e) => panic!("Error creating group from Welcome: {:?}", e),
+        };
+
+
+        // === Alice generates multiple messages ===
+        let MARGIN = 6;
```

Comment:
Maybe i'm missing something, but the constant is inside sender_ratchet and not visible in the context of the integration test. (both the sender_ratchet and the constants)

I'm thinking about moving the boundary tests to a unit test inside the crate, and making both constants public to the crate.

Reply:
I just saw that we already have a place for unit tests for the secret tree here: https://github.com/openmls/openmls/blob/main/openmls/src/tree/tests_and_kats/unit_tests/test_secret_tree.rs

I think the best approach would be to extend that file with your unit tests (using `SenderRatchet` rather than the group).

Reply:
I was already looking into creating a single file with test_sender_ratchet.rs alongside with the test_secret_tree.rs to have the tests specific to the ratchet. 

The idea of the integration test was to indirectly test that secret_for_encryption and secret_for_decryption are producing the same secrets for the generations, and keeping the correct secret order in the "past_secrets" , thus deriving the same ratchet.
 
I can either take this into account in new tests or have both (an out-of-order group message) generic test (without the boundary checks) to avoid the need of public constants outside the crate, testing the whole scenario, whatever is more comfortable for you.

Reply:
Regarding the constants: I created #457 because I think we want the constants to be configurable in the future and then it wouldn't be a problem to access them from outside the crate. For an outside integration test you would have to either make the constants public or duplicate them for now. Integration tests should however go here: https://github.com/openmls/openmls/tree/main/openmls/tests
The other location is just for unit tests.